ENIGMAsystem/SHELL/Platforms/General/UNIXfilemanip.cpp

```diff
@@ -20,48 +20,626 @@
+            {
+                size_t sz2 = sb2.st_size;
+                return ((ret == 1) ||
+                    (sz1 == 0 && sz2 == 0));
+            }
+        }
+    }
+
+    return 0;
 }
 
-int file_rename(string oldname,string newname)
+int directory_exists(std::string dname)
 {
-  return rename(oldname.c_str(),newname.c_str());
+    std::string str(dname);
+
+    if (!str.empty())
+    {
+        while (*str.rbegin() == '/')
+        {
+            str.erase(str.size()-1);
+        }
+    }
+
+    struct stat sb;
+    return (stat((char *)str.c_str(),&sb) == 0 &&
+        S_ISDIR(sb.st_mode));
 }
 
-int file_copy(string fname,string newname)
+int directory_create(std::string dname)
 {
-  return system(("cp "+fname+" "+newname).c_str()); // Hackish, but there's no good implementation on Linux
+    std::string str(dname);
+
+    if (!str.empty())
+    {
+        while (*str.rbegin() == '/')
+        {
+            str.erase(str.size()-1);
+        }
+    }
+
+    struct stat sb;
+    if (stat((char *)str.c_str(),&sb) != 0)
+    {
+        return (mkdir((char *)str.c_str(),0777) == 0);
+    }
+
+    return 0;
 }
 
-int directory_exists(string dname)
+int directory_destroy(std::string dname)
 {
-  struct stat st;
-  return (stat(dname.c_str(),&st) == 0) and (S_ISDIR(st.st_mode));
+    std::string str(dname);
+
+    if (!str.empty())
+    {
+        while (*str.rbegin() == '/')
+        {
+            str.erase(str.size()-1);
+        }
+    }
+
+    struct stat sb;
+    if (stat((char *)str.c_str(),&sb) == 0 &&
+        S_ISDIR(sb.st_mode))
+    {
+        FTS *ftsp = NULL;
+        FTSENT *curr;
+
+        char *files[] = {(char *)str.c_str(),NULL};
+        ftsp = fts_open(files,FTS_NOCHDIR|FTS_PHYSICAL|FTS_XDEV,NULL);
+        if (!ftsp)
+        {
+            return 0;
+            goto finish;
+        }
+
+        while ((curr = fts_read(ftsp)))
+        {
+            switch (curr->fts_info)
+            {
+                case FTS_NS:
+                case FTS_DNR:
+                case FTS_ERR:
+                    return 0;
+                    break;
+
+                case FTS_DC:
+                case FTS_DOT:
+                case FTS_NSOK:
+                    break;
+
+                case FTS_D:
+                    break;
+
+                case FTS_DP:
+                case FTS_F:
+                case FTS_SL:
+                case FTS_SLNONE:
+                case FTS_DEFAULT:
+                    if (remove(curr->fts_accpath) < 0)
+                    {
+                        return 0;
+                    }
+                    break;
+            }
+        }
+
+        finish:
+            if (ftsp)
+            {
+                fts_close(ftsp);
+            }
+
+        return 1;
+    }
+
+    return 0;
 }
 
-int directory_create(string dname) {
-  return mkdir(dname.c_str(),S_IRUSR|S_IWUSR|S_IXUSR);
+int directory_rename(std::string oldname,std::string newname)
+{
+    std::string str1(oldname);
```

Comment:
Once again, this is a ***LOT*** of code to do something as simple as rename a file. Most of the behavior you're defining here is already defined by `rename()`. If you want to have special handling for when the destination file already exists, that's fine, but that's as simple as calling `stat()`. Why are you trying to decompose these paths yourself? If you *absolutely need* that information, please use a call to `filename_name()` rather than re-implementing it yourself in every function. Regardless of what you do, please use more helpers instead of duplicating the code everywhere—it creates a maintenance problem, because someone will go in and make one fix to one function, but won't make the same change in the other dozen functions that are doing the same thing.